### DISCUSSION TITLE
Support linting multiple files from different directories

### DISCUSSION THREAD

**Discussion by flokli** [terraform-linters/tflint#1676] [General]
## Introduction

`tflint` should allow specifying multiple files from different directories to be linted.

This was already proposed once, but ultimately decided against, as `terraform` back then also didn't support formatting multiple files: https://github.com/terraform-linters/tflint/issues/527#issuecomment-866134884

In the meantime, `terraform fmt` gained support for formatting multiple files, from multiple directories: https://github.com/hashicorp/terraform/pull/28191

## Proposal

`tflint` should allow receiving multiple files to be formatted. It could probably still internally bucket them together.

## References

 - https://github.com/numtide/treefmt/blob/main/docs/formatters-spec.md
 - https://numtide.github.io/treefmt/

**Comment by bendrucker** [comment]
You've left out why this is valuable, which is mentioned in the TF thread:

https://github.com/hashicorp/terraform/pull/28191#issuecomment-834723899

> treefmt allows dramatically speeding up repository formatting by only attempting to format files that have changed

That said, `fmt` operates purely on syntax and can operate on individual files. TFLint is not a formatter. TFLint has rules that expect access to the whole module scope, e.g., `terraform_required_providers`.  It is incorrect to call TFLint with a single file though I think it's still supported, perhaps with a warning. You always need to operate on the module (directory) that was changed and not the file.

I'm not convinced it's TFLint's responsibility to take in files and then implicitly step up to their directories and then get the unique set of directories. Especially as we're actively moving away from args and towards conformance with Terraform's module-oriented commands.

**Comment by flokli** [comment]
Hm, would it still make sense to allow multiple directories to be passed, instead of a single one?

That could save spinning up n `tflint` processes, one for each directory...

**Comment by wata727** [comment]
Module scope includes the current directory.

For example, evaluation of `path.cwd` obviously depends on the current directory, so Terraform's semantics do not allow multiple directories to be evaluated simultaneously.

> That could save spinning up n tflint processes, one for each directory...

A solution to this issue is `--recurisve`. Wouldn't this help?

**Reply by bendrucker** [reply]
The idea here would be that you have a repository with thousands of modules and want to execute TFLint only in changed modules. So you can't just recurse, you'd have multiple directories. You can handle this already, but not with a single call to TFLint.
